### ISSUE TITLE
Write tutorial for Cellpose (and maybe create workflow)

### ISSUE THREAD

**Issue by kostrykin** [beatrizserrano/BH2024-project17#15]

**Comment by rmassei** [comment]
Hi @kostrykin I am on it and I am going to test around with the datasets you just posted
(can you assign this to me so I can keep track of it?)

**Comment by rmassei** [comment]
l I think this might be a new issue, but I can open the discussion here. 
Before starting with the training, I would like to open PR to slightly modify the headless Cellpose tool.

**Small fixes to better handle TIFFs and multichannels images:**
1) When processing a TIFF file, cellpose segmentation either: 1) fail since the TIFF is not multichannel or 2) succeeds, but the axes are transposed incorrectly during processing. If I understand correctly [the Python code](https://github.com/bgruening/galaxytools/blob/13c64a37f864ac93fcbdfdd0a149302eaf31df0f/tools/cellpose/cp_segmentation.py#L56C1-L61C80):

```
    print(f"Image shape: {img.shape}")
    # transpose to Ly x Lx x nchann and reshape based on channels
    if img_format.endswith('tiff'):
        img = np.transpose(img, (1, 2, 0))
        img = transforms.reshape(img, channels=channels, chan_first=chan_first)
```

it assumes that the TIFF is always multichannel AND the channel dimension is always the third axis.... which is not always the case. The fast solution is to convert the file to PNG before processing with Cellpose, but ideally the code should be made more flexible to handle TIFF files with different axis orders or single-channel data.

2) The current channel selection assumes RGB ordering, which can be confusing or non-RGB images:

<img width="884" height="297" alt="Image" src="https://github.com/user-attachments/assets/fa13b1cd-614f-40cf-b78b-5f539983158c" />

It might be more intuitive to allow the user to specify the channel index directly (e.g., channel=0, channel=1, etc.) instead of hardcoded RGB mappings.

@kostrykin do you have any thoughts or I am misreading the tool?

**Comment by kostrykin** [comment]
Hi @rmassei and thanks for bringing this up.

TIFFs are very flexible, which requires quite some attention, when dealing with them. Axis can be ordered arbitrarily, and you're absolutely right, the wrapper code currently assumes that channels are the third axis, which is bad. Also, even if a channel axis is present, *and* it is the third axis, there is no guarantee that there are 3 channels included. It might as well be just a single one, which, again, would make the wrapper fail. Coping with that flexibility of TIFF can be nasty.

To cope with that, I have written the [giatools](https://bioconda.github.io/recipes/giatools/README.html) Python module, that brings some auxiliary functions to facilitate reading and writing TIFFs (and images in general), while having a minimal set of dependencies (tifffile and pandas, I think). The module reads a TIFF with arbitrary axes, in arbitrary order, and then yields a NumPy array that comes with a *normalized* set of axes, in a *normalized* order ([docs](https://giatools.readthedocs.io/en/latest/giatools.image.html#giatools.image.Image.read)), without any copying of data, so it's efficient. By default, the normalized axes are `QTZYXC`, i.e. channels will be the last axis. So if you read an RGB image with giatools as NumPy array `arr`, it will have `arr.shape[-1] == 3`, and if you read an image without channels or with only a single channel, it will have `arr.shape[-1] == 1`. 

AFAIK Cellpose is only capable of segmenting a single channel at once, isn't it? So the whole channel extraction procedure is just for convenience, at the heart, the module works with single-channel data, doesn't it? In that case, personally, I would suggest to simplify the wrapper so it will only accept single-channel images. If you would need to perform segmentation on a specific channel of a multi-channel image, you could still extract it using a designated tool (e.g., [this](https://usegalaxy.eu/root?tool_id=toolshed.g2.bx.psu.edu/repos/imgteam/split_image/ip_split_image/2.2.3+galaxy1)). I think this is reasonable, given that working with channels is more complex than one would assume at first, as described above.

Let me know if you want to discuss further. Also, if you think that whatever solution you take is subpar, feel free to open a new issue and write down what should be improved, so we can do that later.

**Comment by rmassei** [comment]
Thanks for the reply!
I think with the existing setting is it possible to segment the nuclei and perform the cytoplasm segmentation in one go which might be a useful feature.
However, I did many attempts with multichannel TIFF (e.g. split the channels and then re-concatenate them again on the third axis, or even, just for testing purposes, converts a single-channel to multi-channel image by repeating the channel multiple times) and it seems that the tool is interpreting the axes order always wrong. 
However, if single channel is extracted from the TIFF with the tool you posted then everything goes fine. 
For sake of simplicity, I would agree to change the wrapper to accept single-channels images. I will open an issue for this!
